### PR TITLE
Fix Docker runtimes not stopping

### PR DESCRIPTION
**Pull request by tofarr**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below
no changelog

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This PR fixes an issue where Docker runtime containers were not being properly stopped when the application shuts down. The changes include:

### Changes

1. Enhanced Shutdown Listener System:
   - Added a new shutdown listener mechanism in `shutdown_listener.py` that works better with Starlette/Uvicorn shutdown signals
   - Replaced the `atexit` handler with the new shutdown listener system
   - Added UUID-based listener registration and removal functionality

2. Docker Runtime Improvements:
   - Removed global `_atexit_registered` flag and `stop_all_runtime_containers` function
   - Added class-level `_shutdown_listener_id` to track the shutdown listener
   - Containers are now properly stopped when the application shuts down

3. Added Comprehensive Tests:
   - New test file `test_shutdown_listener.py` with thorough test coverage
   - Tests for listener registration, removal, and shutdown behavior
   - Tests for edge cases like removing listeners during shutdown

### Technical Details
- The new shutdown listener system uses a dictionary to track listeners by UUID
- Signal handlers are properly chained to preserve existing handlers
- Listeners are called exactly once during shutdown
- Added safeguards against recursive shutdown calls

This change ensures that Docker containers are properly cleaned up when the application exits, preventing orphaned containers from consuming system resources.

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:26f53e2-nikolaik   --name openhands-app-26f53e2   docker.all-hands.dev/all-hands-ai/openhands:26f53e2
```

### CODE REVIEW

File: openhands/runtime/impl/docker/docker_runtime.py [hunk 1]
```diff
@@ -66,10 +62,8 @@ def __init__(
         attach_to_existing: bool = False,
         headless_mode: bool = True,
     ):
-        global _atexit_registered
-        if not _atexit_registered:
-            _atexit_registered = True
-            atexit.register(stop_all_runtime_containers)
+        if not self._shutdown_listener_id:
```
[the remark this answers is not in the document]
Comment (tofarr) [reply openhands/runtime/impl/docker/docker_runtime.py:65]: Python classes will search for attributes in self, then self.__class__ by default anyway:
image

[the remark this answers is not in the document]
Comment (tofarr) [reply openhands/runtime/impl/docker/docker_runtime.py]: I've updated to use the class name for clarity that it's a class level variable

[the remark this answers is not in the document]
Comment (rbren) [reply openhands/runtime/impl/docker/docker_runtime.py]: 👍 yeah I was just confused between the two different references. Python is weird sometimes

**Review by rbren — APPROVED** [review]
